src/analysis/calculus/taylor.lean

```diff
@@ -0,0 +1,380 @@
+  (hs' : s' ∈ 𝓝[s] y) (hy : y ∈ s') (h : s' ⊆ s)
+  (hf : cont_diff_on ℝ n f s)
+  (hf' : differentiable_on ℝ (iterated_deriv_within n f s) s') :
+  has_deriv_within_at (λ t, taylor_within_eval f n s t x)
+    (((n.factorial : ℝ)⁻¹ * (x - y)^n) • (iterated_deriv_within (n+1) f s y)) s' y :=
+begin
+  induction n with k hk,
+  { simp only [taylor_within_zero_eval, nat.factorial_zero, nat.cast_one, inv_one, pow_zero,
+      mul_one, zero_add, one_smul],
+    simp only [iterated_deriv_within_zero] at hf',
+    rw iterated_deriv_within_one hs_unique (h hy),
+    refine has_deriv_within_at.mono _ h,
+    refine differentiable_within_at.has_deriv_within_at _,
+    exact (hf' y hy).antimono h hs' },
+  simp_rw [nat.add_succ, taylor_within_eval_succ],
+  simp only [add_zero, nat.factorial_succ, nat.cast_mul, nat.cast_add, nat.cast_one],
+  have hdiff : differentiable_on ℝ (iterated_deriv_within k f s) s' :=
+  begin
+    have coe_lt_succ : (k : with_top ℕ) < k.succ :=
+    by { rw [with_top.coe_lt_coe], exact lt_add_one k },
+    refine differentiable_on.mono _ h,
+    exact hf.differentiable_on_iterated_deriv_within coe_lt_succ hs_unique,
+  end,
+  specialize hk (cont_diff_on.of_succ hf) hdiff,
+  convert hk.add (taylor_coeff_within_has_deriv_within_at hs'_unique hs' hy h hf'),
+  exact (add_sub_cancel'_right _ _).symm,
+end
+
+/-- Calculate the derivative of the Taylor polynomial with respect to `x₀`.
+
+Version for open intervals -/
+lemma taylor_within_eval_has_deriv_at_Ioo {f : ℝ → E} {a b t : ℝ} (x : ℝ) {n : ℕ}
+  (hx : a < b) (ht : t ∈ set.Ioo a b)
+  (hf : cont_diff_on ℝ n f (set.Icc a b))
+  (hf' : differentiable_on ℝ (iterated_deriv_within n f (set.Icc a b)) (set.Ioo a b)) :
+  has_deriv_at (λ y, taylor_within_eval f n (set.Icc a b) y x)
+    (((n.factorial : ℝ)⁻¹ * (x - t)^n) • (iterated_deriv_within (n+1) f (set.Icc a b) t)) t :=
+begin
+  have h_nhds := is_open.mem_nhds is_open_Ioo ht,
+  exact (taylor_within_eval_has_deriv_within_at (unique_diff_within_at_Ioo ht)
+    (unique_diff_on_Icc hx) (nhds_within_le_nhds h_nhds) ht set.Ioo_subset_Icc_self hf hf')
+    .has_deriv_at h_nhds,
+end
+
+/-- Calculate the derivative of the Taylor polynomial with respect to `x₀`.
+
+Version for closed intervals -/
+lemma taylor_within_eval_has_deriv_within_at_Icc {f : ℝ → E} {a b t : ℝ} (x : ℝ) {n : ℕ}
+  (hx : a < b) (ht : t ∈ set.Icc a b) (hf : cont_diff_on ℝ n f (set.Icc a b))
+  (hf' : differentiable_on ℝ (iterated_deriv_within n f (set.Icc a b)) (set.Icc a b)) :
+  has_deriv_within_at (λ y, taylor_within_eval f n (set.Icc a b) y x)
+    (((n.factorial : ℝ)⁻¹ * (x - t)^n) • (iterated_deriv_within (n+1) f (set.Icc a b) t))
+    (set.Icc a b) t :=
+taylor_within_eval_has_deriv_within_at (unique_diff_on_Icc hx t ht) (unique_diff_on_Icc hx)
+  self_mem_nhds_within ht rfl.subset hf hf'
+
+/-! ### Taylor's theorem with mean value type remainder estimate -/
+
+/-- **Taylor's theorem** with the general mean value form of the remainder. -/
```

Comment:
Can you explain what the theorem says, in the docstring? When a theorem has a lot of assumptions like that, it is often pretty hard to parse, and a good docstring can help a lot! Same thing below.